src/daemon/engine/svc/node/exec_cmd_service.cpp

```diff
@@ -0,0 +1,265 @@
+        {
+            logger().trace("ExecCmdSvc::Fsm (id={}).", id_);
+
+            channel_.subscribe([this](const auto& event_var) {
+                //
+                cetl::visit([this](const auto& event) { handleEvent(event); }, event_var);
+            });
+        }
+
+        ~Fsm()
+        {
+            logger().trace("ExecCmdSvc::~Fsm (id={}).", id_);
+        }
+
+        Fsm(const Fsm&)                = delete;
+        Fsm(Fsm&&) noexcept            = delete;
+        Fsm& operator=(const Fsm&)     = delete;
+        Fsm& operator=(Fsm&&) noexcept = delete;
+
+        void start(const Spec::Request& request)
+        {
+            logger().trace("ExecCmdSvc::Fsm::start (fsm_id={}).", id_);
+
+            // Immediately complete if there are no nodes to execute the command on.
+            //
+            if (request.node_ids.empty())
+            {
+                complete(0);
+                return;
+            }
+
+            // It's ok to have duplicates in the request -
+            // we just ignore duplicates, and work with unique ones.
+            const SetOfNodeIds unique_node_ids{request.node_ids.begin(), request.node_ids.end()};
+
+            const auto deadline = service_.context_.executor.now() + std::chrono::microseconds{request.timeout_us};
+            const CyphalExecCmdSvc::Request cy_request{request.payload.command, request.payload.parameter, &memory()};
+            for (const auto node_id : unique_node_ids)
+            {
+                if (const auto err = makeCyphalSvcCallFor(deadline, node_id, cy_request))
+                {
+                    complete(err);
+                    return;
+                }
+            }
+        }
+
+    private:
+        using SetOfNodeIds         = std::unordered_set<std::uint16_t>;
+        using CyphalExecCmdSvc     = uavcan::node::ExecuteCommand_1_3;
+        using CyphalSvcClient      = libcyphal::presentation::ServiceClient<CyphalExecCmdSvc>;
+        using CyphalPromise        = libcyphal::presentation::ResponsePromise<CyphalExecCmdSvc::Response>;
+        using CyphalPromiseFailure = libcyphal::presentation::ResponsePromiseFailure;
+
+        struct CyNodeOp
+        {
+            CyphalSvcClient client;
+            CyphalPromise   promise;
+        };
+
+        common::Logger& logger() const
+        {
+            return *service_.logger_;
+        }
+
+        cetl::pmr::memory_resource& memory() const
+        {
+            return service_.context_.memory;
+        }
+
+        // We are not interested in handling these events.
+        static void handleEvent(const Channel::Connected&) {}
+        static void handleEvent(const Channel::Input&) {}
+
+        void handleEvent(const Channel::Completed& completed)
+        {
+            logger().debug("ExecCmdSvc::Fsm::handleEvent({}) (id={}).", completed, id_);
+            complete(ECANCELED);
+        }
+
+        int makeCyphalSvcCallFor(const libcyphal::TimePoint       deadline,
```

Comment:
Please provide a specification for this state machine in a format of your choice. asciiflow in comments is fine if it's trivial otherwise a draw.io diagram would be in order etc. It's impossible to asses the correctness of a state machine without its specification in hand.

Reply:
ok, will do.